### PR TITLE
remove (the never released) prefix matching in tctl

### PR DESCRIPTION
**Pull request by GavinFrazar**
This PR replaces the prefix matching in `tctl` with exact discovered name match as a fallback when no resource full name matches the name given by a user.

changelog: none

This is for consistency with other commands where we decided to do the same.

### CODE REVIEW

**Comment by GavinFrazar** [comment]
> This may not be a "breaking" change, but it is a behavior change and I don't think it's something that should be released in a patch release.

I have been holding back on actually backporting these UX PRs. The `tctl` prefix matching PR https://github.com/gravitational/teleport/pull/30081 was never backported into v13 for release.

**Comment by zmb3** [comment]
Ok, I'm confused why there's a backport label for v13 then.

**Comment by GavinFrazar** [comment]
> Ok, I'm confused why there's a backport label for v13 then.

because it will be backported to v13 - I only held off backporting until after the test plan was done, but I kept all the PRs labelled properly for when I actually do that... which is today/tomorrow

**Comment by zmb3** [comment]
If you backport it to v13 it's going to land in a minor or patch release. While unlikely, it's possible that this change breaks valid scripts that depend on the current behavior.

Is that something we're okay with? I don't have context on the reasons for this change but on the surface it seems backwards incompatible.

**Comment by smallinsky** [comment]
@zmb3 
We have agreed in RFD that this resource name collision is breaking change but in order to minimalize the UX impact 
we have decided to backport UX tctl,tsh change to v13 to to make the transition smother so if the teleport cluster was upgraded to v14 but a user client is still using the newest version of tsh v13 (with backported alignment) they flow will not be affected:  
https://github.com/gravitational/teleport/blob/master/rfd/0129-discovery-name-templating.md#backward-compatibility

File: tool/tctl/common/resource_command_test.go [hunk 1]
```diff
@@ -102,9 +119,12 @@ func TestDatabaseServerResource(t *testing.T) {
 	}
 
 	db1, err := types.NewDatabaseV3(types.Metadata{
-		Name:        "example",
+		Name:        "example-rds-us-west-1",
 		Description: "Example MySQL",
-		Labels:      map[string]string{types.OriginLabel: types.OriginConfigFile},
+		Labels: map[string]string{
+			types.OriginLabel:         types.OriginConfigFile,
```
Comment (greedy52) [inline tool/tctl/common/resource_command_test.go:125]: nit:
```suggestion
			types.OriginLabel:         types.OriginCloud,
```
same for `"example-rds-us-west-2"`

**Comment by zmb3** [comment]
Thanks for clarifying @smallinsky!

File: tool/tctl/common/resource_command_test.go [hunk 2]
```diff
@@ -902,54 +920,68 @@ func (test *dynamicResourceTest[T]) run(t *testing.T) {
...
 
-	// Test prefix name behavior.
-	// Removing multiple resources ("foo" and "foo-bar-baz")by prefix name is an error.
-	_, err = runResourceCommand(t, fileConfig, []string{"rm", fmt.Sprintf("%v/%v", test.kind, "f")})
+	// Test discovered name behavior.
+	// Fetching multiple resources ("foo-bar-1" and "foo-bar-2") by discovered name is ok.
+	buf, err = runResourceCommand(t, fileConfig, []string{"get", fmt.Sprintf("%v/%v", test.kind, "foo-bar"), "--format=json"})
+	require.NoError(t, err)
+	out = nil
+	mustDecodeJSON(t, buf, &out)
```
Comment (greedy52) [inline tool/tctl/common/resource_command_test.go:964]: nit: maybe make a new helper to do the reset then decode

**Review by greedy52 — APPROVED** [review]
LGTM.

Not directly related to this change but I am surprised that `getCollection` in general gets all resources then filter locally. I guess that's for another day.

**Review by smallinsky — APPROVED** [review]

File: tool/tctl/common/resource_command_test.go (hunk 1 above)
Comment (GavinFrazar) [reply tool/tctl/common/resource_command_test.go:125]: it gets set by teleport init to config file

File: tool/tctl/common/resource_command_test.go (hunk 2 above)
Comment (GavinFrazar) [reply tool/tctl/common/resource_command_test.go]: Refactored the yaml/json decoders to do exactly that: 3adaf7d330c7b0188d8275bbed05b437b5710f6c

I realized while reworking the tests that when the var is already decoded into it should be reset before reuse, so this prevents that mistake from happening again

**Comment by GavinFrazar** [comment]
@zmb3 this PR touched a lot of tests and makes the flaky detector timeout again :(

```
- tool/tctl/common/lock_command_test.go:
  ~TestLocks (TestLocks)

- tool/tctl/common/resource_command_test.go:
  +TestFilterByNameOrDiscoveredName (TestFilterByNameOrDiscoveredName)
  ~TestDatabaseServerResource (TestDatabaseServerResource)
  ~TestDatabaseServiceResource (TestDatabaseServiceResource)
  ~TestIntegrationResource (TestIntegrationResource)
  ~TestCreateLock (TestCreateLock)
  ~TestCreateClusterAuthPreference_WithSupportForSecondFactorWithoutQuotes (TestCreateClusterAuthPreference_WithSupportForSecondFactorWithoutQuotes)
  ~TestCreateSAMLIdPServiceProvider (TestCreateSAMLIdPServiceProvider)
  ~TestDatabaseResource (TestDatabaseResource)
  ~TestKubeClusterResource (TestKubeClusterResource)
  ~TestAppResource (TestAppResource)
  ~TestGetOneResourceNameToDelete (TestGetOneResourceNameToDelete)
  ~TestFormatAmbiguousDeleteMessage (TestFormatAmbiguousDeleteMessage)

- tool/tctl/common/token_command_test.go:
  ~TestTokens (TestTokens)
```

**Pull request merged**